cmd/krew/cmd/root.go

```diff
@@ -25,18 +25,18 @@ import (
 	"github.com/spf13/pflag"
 	"k8s.io/klog"
 
-	"sigs.k8s.io/krew/internal/assertion"
 	"sigs.k8s.io/krew/internal/environment"
 	"sigs.k8s.io/krew/internal/gitutil"
 	"sigs.k8s.io/krew/internal/installation"
 	"sigs.k8s.io/krew/internal/installation/receipt"
 	"sigs.k8s.io/krew/internal/receiptsmigration"
+	"sigs.k8s.io/krew/internal/updatecheck"
 	"sigs.k8s.io/krew/pkg/constants"
 )
 
 var (
-	paths         environment.Paths // krew paths used by the process
-	notifications = make(chan string)
+	paths               environment.Paths // krew paths used by the process
+	upgradeNotification = ""
```

Comment:
I'm thinking something like

```
latestVersionAvailable string
```

is better here coz then we can compare it with internal version at the end & print the msg. let's not prep the msg so early.

Reply:
My rationale was that it improves the testability of the message. But I agree that it's not very pretty right now.

Reply:
Also, how do we ensure previously started check is completed when compared in the current implementation ?

Additionally, isn't there a race in setting and reading the last version?

Reply:
Hmm, I think the "race condition" question comes down to _"is setting a slice reference atomic in go?"_.

As far as I know, this is just adjusting the pointer from NULL to the slice struct. So if reference assigments in go are atomic (assuming single-word address space?) we should be good.

Reply:
> how do we ensure previously started check is completed

that should be simply available as checking if `latestVersionAvailable == ""`  or not.

Reply:
The second question might be irrelevant when the first is updated or resolved.

I think I wasn't clear for the first question, what we do know is:

* start a goroutine to prepare an update notification
* run the main command
* check if we have a non-zero update notification and print -- but we don't know yet if it's completed. Since they aren't communicating, there is a race between writing and reading.

Let me know if I'm missing something.

Reply:
We don't "wait" for update notification to complete, by design. (read #480)

I can't see the race condition. What does it lead to (corrupt pointer reading?)? If the result is not written by the time we read it, we don't wait for it intentionally.

Reply:
Sorry, I should have checked #480 first. 

If this is by design, it's fine.